# Importing Liabraries

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import matplotlib.axes as ax

In [2]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

### About Data file


In [3]:
features = {'A':'area',
            'P':'perimeter',
            'C':'compactness',
            'LK':'length',
            'WK':'width',
            'A_Coef':'asymmetry',
            'LKG':'length_kernel_groove'}

## Loading Data

In [4]:
data = pd.read_csv('Seed_Data.csv')

In [5]:
data.rename(columns=features)

area  perimeter  compactness  length  width  asymmetry  \
0    15.26      14.84       0.8710   5.763  3.312      2.221   
1    14.88      14.57       0.8811   5.554  3.333      1.018   
2    14.29      14.09       0.9050   5.291  3.337      2.699   
3    13.84      13.94       0.8955   5.324  3.379      2.259   
4    16.14      14.99       0.9034   5.658  3.562      1.355   
..     ...        ...          ...     ...    ...        ...   
205  12.19      13.20       0.8783   5.137  2.981      3.631   
206  11.23      12.88       0.8511   5.140  2.795      4.325   
207  13.20      13.66       0.8883   5.236  3.232      8.315   
208  11.84      13.21       0.8521   5.175  2.836      3.598   
209  12.30      13.34       0.8684   5.243  2.974      5.637   

     length_kernel_groove  target  
0                   5.220       0  
1                   4.956       0  
2                   4.825       0  
3                   4.805       0  
4                   5.175       0  
..                    ...     ...  
205                 4.870       2  
206                 5.003       2  
207                 5.056       2  
208                 5.044       2  
209                 5.063       2  

[210 rows x 8 columns]

In [6]:
data.head()

A      P       C     LK     WK  A_Coef    LKG  target
0  15.26  14.84  0.8710  5.763  3.312   2.221  5.220       0
1  14.88  14.57  0.8811  5.554  3.333   1.018  4.956       0
2  14.29  14.09  0.9050  5.291  3.337   2.699  4.825       0
3  13.84  13.94  0.8955  5.324  3.379   2.259  4.805       0
4  16.14  14.99  0.9034  5.658  3.562   1.355  5.175       0

<font color = 'red'> <b> Depending on parameters we need to Predict TARGET column

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210 entries, 0 to 209
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       210 non-null    float64
 1   P       210 non-null    float64
 2   C       210 non-null    float64
 3   LK      210 non-null    float64
 4   WK      210 non-null    float64
 5   A_Coef  210 non-null    float64
 6   LKG     210 non-null    float64
 7   target  210 non-null    int64  
dtypes: float64(7), int64(1)
memory usage: 13.2 KB


<font color = 'red'> <b> No column have NULL value all have 210 entries

## Counting Unique Values

In [8]:
data['target'].unique()

array([0, 1, 2], dtype=int64)

In [9]:
data['target'].value_counts()

0    70
1    70
2    70
Name: target, dtype: int64

<font color = 'red'> <b> Have 3 unique values 0,1,2 and all of equal quanity Perfect_Dataset

## Splitting Data

In [10]:
x = data.drop(data.columns[[7]],axis = 1)
y = data['target']

In [11]:
x.head()

A      P       C     LK     WK  A_Coef    LKG
0  15.26  14.84  0.8710  5.763  3.312   2.221  5.220
1  14.88  14.57  0.8811  5.554  3.333   1.018  4.956
2  14.29  14.09  0.9050  5.291  3.337   2.699  4.825
3  13.84  13.94  0.8955  5.324  3.379   2.259  4.805
4  16.14  14.99  0.9034  5.658  3.562   1.355  5.175

In [16]:
#now splitting x and y in 2 form 

from sklearn.model_selection import train_test_split

xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=0.25, random_state=0)

In [17]:
print(xtrain.shape)
print(ytrain.shape)

print(xtest.shape)
print(ytest.shape)

(52, 7)
(52,)
(158, 7)
(158,)


<font color = 'red'> <b> From x.head we can see our data have diffrent distribution so we need to do DataScaling

# Feature Scaling

In [18]:
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()

In [19]:
xtrain = sc_x.fit_transform(xtrain)

xtest = sc_x.transform(xtest)
#here not using fit because we want to fit accordingly to prefiusly fitted data

# <font color = ' green'><b> K - Nearest Neighbour (KNN)

In [20]:
from sklearn.neighbors import KNeighborsClassifier

In [54]:
knn_classifer = KNeighborsClassifier(n_neighbors = 3 , p = 2)

# here p= 2 means equilidean distance 1 for malhaten distance

knn_classifer.fit(xtrain,ytrain)

KNeighborsClassifier(n_neighbors=3)

## Predicting test_inputs

In [55]:
y_predict = knn_classifer.predict(xtest)

## Result and Accuray

In [56]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
cm = confusion_matrix(ytest,y_predict)
cm

array([[40,  2, 14],
       [ 4, 50,  0],
       [ 0,  0, 48]], dtype=int64)

In [57]:
accuracy_score(ytest,y_predict)

0.8734177215189873

<font color = 'red'> it showes that model is 87.34 percent accurate 